### PR TITLE
Add softmax_focal_loss() to allow multi-class focal loss

### PR DESCRIPTION
**Pull request by dhruvbird**
In image segmentation tasks, focal loss is useful when trying to classify an image pixel as one of N classes. Unfortunately, `sigmoid_focal_loss()` isn't useful in such cases. I found that other have been asking for this as well here https://github.com/pytorch/vision/issues/3250 so I decided to submit a PR for the same.

Putting up this PR to get feedback about whether this sounds okay.

### CODE REVIEW

File: torchvision/ops/focal_loss.py [hunk 1]
```diff
@@ -56,3 +57,95 @@ def sigmoid_focal_loss(
...
+    #
+    # Why does this work?
+    # Since this is a multi-class setting, only one class is active. The
+    # probability of that class is 1, and the rest are all 0.
+    #
+    # Cross Entropy Loss computes:
+    # pt = softmax(...)
+    # loss = -1.0 * log(pt)
+    #
+    # Hence, exp(-loss) == pt
```
Comment (rehno-lindeque) [inline torchvision/ops/focal_loss.py:134]: Sadly I don't think this works anymore:

`loss = -weight_t * log(pt)`
`exp(-loss) == exp(weight_t * log(pt)`
`exp(-loss) == exp(log(pt ** weight_t))` due to the power rule of logarithms
`exp(-loss) == pt ** weight_t`


From what I've seen, I think this will instead require

```python
p = softmax(...)
ce_loss = nll_loss(p, weight=weight)
pt = p[:, targets]
```


Hence, 
`((1 - pt) ** gamma) * ce_loss` 
== `((1 - pt) ** gamma) * (-weight_t * log(pt))`
== `-weight_t * ((1 - pt) ** gamma) * log(pt)`

Comment (dhruvbird) [reply torchvision/ops/focal_loss.py:134]: @rehno-lindeque Please could you check the file changes and see if the code looks okay - I think the comment may suond misleading. The code is computing cross entropy loss twice, once without the weight and once with the weights, so that the one without can be exponentiated to compute the probability.

Comment (rehno-lindeque) [reply torchvision/ops/focal_loss.py:134]: Oh I did indeed miss that, you are right the code is functionally correct. Sorry about that.

I would still humbly suggest breaking down the cross_entropy into softmax followed by nll_loss as I described for performance reasons though.

That is, calculating cross entropy twice might be rather expensive, especially if the tensors are images. It's only necessary to calculate it once if you take `pt` directly from the softmax result.
